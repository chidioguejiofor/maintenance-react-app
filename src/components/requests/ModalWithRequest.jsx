
import React from 'react';
import { Modal, Button, Container, Header, Label,
  Image, Icon } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import { getColor, getIconType,
  capitalizeWord } from '../../helpers/requestHelper';

const ModalWithRequest = ({
  title, description, location, status, date, image
}) => (
  <span id="modal-with-request">
    <Modal

      size="small"
      trigger={(
        <Button
          size="small"
          content="Info"
          icon="info"
        />
      )}
    >
      <Modal.Header>{title.toUpperCase()}</Modal.Header>
      <Modal.Content>
        <Container>

          <Image centered src={image} size="big" />
          <div className="request">
            <Header size="large">Created On </Header>
            <p>{`${new Date(date).toDateString()}`}</p>
            <Header size="large">Description</Header>
            <p>{description}</p>
            <Header size="large">Location </Header>
            <p>{location}</p>
            <Label size="large">
              <Icon
                name={getIconType(status)}
                color={getColor(status)}
              />
              {capitalizeWord(status)}
            </Label>
          </div>

        </Container>
      </Modal.Content>
    </Modal>

  </span>
);


ModalWithRequest.propTypes = {
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  location: PropTypes.string.isRequired,
  status: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,

};


export default ModalWithRequest;
