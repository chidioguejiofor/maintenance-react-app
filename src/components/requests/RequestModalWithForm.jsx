
import React from 'react';
import { Modal, Button } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import Form from '../Form';

const RequestModalWithForm = ({
  onSubmit,
  onChange,
  onClick,
  title, description, location, image,
  update,
  buttonLoading,
  success,
  loading,
  messageTitle,
  errorMessages,
  open,

}) => {
  const options = {
    iconUrl: image || 'https://image.flaticon.com/icons/svg/263/263128.svg',
    formTitle: update ? 'Update Request' : 'Create Request',
    handleSubmit: onSubmit,
    success,
    messageTitle,
    loading,
    handleChange: onChange,
    errorMessages,
    submitLoading: buttonLoading,
    inputs: [{
      name: 'title',
      placeholder: 'Title',
      value: title,
      type: 'text'
    },
    {
      name: 'description',
      placeholder: 'Description',
      value: description,
      type: 'text'
    },
    {
      name: 'location',
      placeholder: 'Location',
      value: location,
      type: 'text'
    },
    {
      name: 'image',
      src: image,
      type: 'file'
    }],
  };


  return (
    <span className="request-button">
      <Modal
        trigger={(
          <Button
            className={update ? 'update-btn' : 'create-btn'}
            color={update ? 'orange' : 'green'}
            size="small"
            content={update ? 'Update' : 'Create'}
            onClick={onClick}
          />
        )}
        open={open}
      >

        <Modal.Content>
          <Form {...options} />
        </Modal.Content>
      </Modal>

    </span>
  );
};

RequestModalWithForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  onClick: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  location: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  update: PropTypes.bool.isRequired,
  buttonLoading: PropTypes.bool.isRequired,
  success: PropTypes.bool.isRequired,
  loading: PropTypes.bool.isRequired,
  messageTitle: PropTypes.string.isRequired,
  errorMessages: PropTypes.arrayOf(PropTypes.string).isRequired,
  open: PropTypes.bool.isRequired,
};

export default RequestModalWithForm;
