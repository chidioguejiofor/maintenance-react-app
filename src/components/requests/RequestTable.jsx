import { Table, Container, Label, Icon } from 'semantic-ui-react';
import React from 'react';
import PropTypes from 'prop-types';
import RequestModalWithForm from './RequestModalWithForm';
import ModalWithRequest from './ModalWithRequest';
import AdminButtons from './AdminButtons';
import { getColor, getIconType,
  capitalizeWord } from '../../helpers/requestHelper';


const RequestTable = ({
  onCreateRequestChange, requests, onSubmit, currentRequest,
  onUpdateClick, onCreateClick, buttonLoading, editedRequest,
  openModals, admin
}) => (
  <div id="request-table">
    <Container>
      <Table singleLine>
        <Table.Header fullWidth>
          <Table.Row>
            <Table.HeaderCell colSpan="12" textAlign="center">
              <span className="table-title">
                YOUR REQUESTS
                <RequestModalWithForm
                  {...currentRequest}
                  onChange={onCreateRequestChange}
                  onClick={onCreateClick}
                  buttonLoading={buttonLoading}
                  onSubmit={onSubmit}
                  success={editedRequest.success}
                  loading={editedRequest.isLoading}
                  messageTitle={editedRequest.message}
                  errorMessages={editedRequest.errors.errorMessages}
                  open={openModals}
                />

              </span>

            </Table.HeaderCell>
          </Table.Row>
        </Table.Header>
        <Table.Body>
          {requests.map(request => (
            <Table.Row key={request.id}>

              <Table.Cell colSpan="6">
                <div className="request-info">

                  <div className="title">
                    {capitalizeWord(request.title) }
                  </div>
                  <div className="description">
                    {request.description}

                  </div>
                  <div>
                  Created On:
                    {` ${new Date(request.date).toDateString()}`}
                  </div>
                  <div className="label-update-group">

                    <ModalWithRequest {...request} />
                    {!admin && request.status === 'pending' ? (
                      <RequestModalWithForm
                        {...currentRequest}
                        onSubmit={() => onSubmit('update')}
                        onChange={onCreateRequestChange}
                        onClick={() => onUpdateClick(request)}
                        update
                        buttonLoading={buttonLoading}
                        success={editedRequest.success}
                        loading={editedRequest.isLoading}
                        messageTitle={editedRequest.message}
                        errorMessages={editedRequest.errors.errorMessages}
                        open={openModals}
                      />
                    ) : null}

                    {admin ? (
                      <AdminButtons
                        status={request.status}
                        onClick={onSubmit}
                        requestId={request.id}
                      />
                    ) : ''}
                    <Label size="large" basic>
                      <Icon
                        name={getIconType(request.status)}
                        color={getColor(request.status)}
                      />
                      {capitalizeWord(request.status)}
                    </Label>
                  </div>
                </div>
              </Table.Cell>
            </Table.Row>


          ))}
        </Table.Body>


      </Table>
    </Container>

  </div>

);

RequestTable.propTypes = {
  onCreateRequestChange: PropTypes.func.isRequired,
  requests: PropTypes.arrayOf(PropTypes.string).isRequired,
  onSubmit: PropTypes.func.isRequired,
  currentRequest: PropTypes.objectOf(PropTypes.string).isRequired,
  onUpdateClick: PropTypes.func.isRequired,
  onCreateClick: PropTypes.func.isRequired,
  buttonLoading: PropTypes.bool,
  editedRequest: PropTypes.objectOf(PropTypes.any).isRequired,
  openModals: PropTypes.bool.isRequired,
};

RequestTable.defaultProps = {
  buttonLoading: false,
};
export default RequestTable;
