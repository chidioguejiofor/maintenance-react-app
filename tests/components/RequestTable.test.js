import React from 'react';
import { shallow } from 'enzyme';
import { Table } from 'semantic-ui-react';
import chai from 'chai';
import { requests as sampleRequests } from '../config/mockData';
import RequestTable from
  '../../src/components/requests/RequestTable';
import ModalWithRequest from
  '../../src/components/requests/ModalWithRequest';

import RequestModalWithForm from
  '../../src/components/requests/RequestModalWithForm';

const sampleProps = {
  editedRequest: {
    success: false,
    isLoading: false,
    message: '',
    errors: {},
  },
  requests: []
};
const getShallowObj = (props = sampleProps) => shallow(
  <RequestTable {...props} />
);
describe('Testing the RequestTable Component', () => {
  describe('the content it renders', () => {
    it('should render a Table', () => {
      const shallowObj = getShallowObj();
      chai.expect(shallowObj.find(Table).length)
        .to.equal(1);
    });
    it('should match snapshot', () => {
      const shallowObj = getShallowObj();
      expect(shallowObj).toMatchSnapshot();
    });
  });

  describe('when the requests array is contains values', () => {
    it('should render a Table.Rows based on the number of requests', () => {
      const shallowObj = getShallowObj({ ...sampleProps, requests: sampleRequests });

      chai.expect(shallowObj
        .find(Table.Row).length)
        .to.be.gte(sampleRequests.length);
      chai.expect(shallowObj
        .find(ModalWithRequest).length)
        .to.equal(sampleRequests.length);
    });
    it('should render the RequestModalWithForm with a submit handler for each request', () => {
      const submitSpy = jest.fn();
      const clickSpy = jest.fn();
      const shallowObj = getShallowObj({
        ...sampleProps,
        requests: sampleRequests,
        onSubmit: submitSpy,
        onClick: clickSpy
      });
      const lastRequestModalProps = shallowObj
        .find(RequestModalWithForm)
        .last().props();
      const firstRequestModalProps = shallowObj
        .find(RequestModalWithForm)
        .last().props();
      lastRequestModalProps.onSubmit();

      expect(submitSpy)
        .toHaveBeenNthCalledWith(1, 'update');
      firstRequestModalProps.onSubmit();
      expect(submitSpy)
        .toHaveBeenCalledTimes(2);
    });

    it('should render the RequestModalWithForm with a click handler for each request', () => {
      const clickSpy = jest.fn();
      const shallowObj = getShallowObj({
        ...sampleProps,
        requests: sampleRequests,
        onUpdateClick: clickSpy
      });
      shallowObj
        .find(RequestModalWithForm)
        .last().simulate('click');
      expect(clickSpy)
        .toHaveBeenCalledTimes(1);
    });
  });
});
