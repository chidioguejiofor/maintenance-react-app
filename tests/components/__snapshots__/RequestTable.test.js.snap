// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the RequestTable Component the content it renders should match snapshot 1`] = `
<div
  id="request-table"
>
  <Container>
    <Table
      as="table"
      singleLine={true}
    >
      <TableHeader
        as="thead"
        fullWidth={true}
      >
        <TableRow
          as="tr"
          cellAs="td"
        >
          <TableHeaderCell
            as="th"
            colSpan="12"
            textAlign="center"
          >
            <span
              className="table-title"
            >
              YOUR REQUESTS
              <RequestModalWithForm
                buttonLoading={false}
                id="create-modal-form"
                loading={false}
                messageTitle=""
                onSubmit={[Function]}
                success={false}
              />
            </span>
          </TableHeaderCell>
        </TableRow>
      </TableHeader>
      <TableBody
        as="tbody"
      />
    </Table>
  </Container>
</div>
`;
